### PR TITLE
Add index name to  cluster block exception

### LINKED ISSUE #40870: `cluster_block_exception` should contain the read-only index name
**Elasticsearch version**: `Version: 7.0.0-rc2, Build: default/tar/7e5c520/2019-04-01T00:56:38.137034Z, JVM: 12`

**JVM version**: 
``` text
openjdk version "12" 2019-03-19
OpenJDK Runtime Environment (build 12+33)
OpenJDK 64-Bit Server VM (build 12+33, mixed mode, sharing)
```

**OS version**: `Darwin MacBook-Pro.localdomain 18.2.0 Darwin Kernel Version 18.2.0: Thu Dec 20 20:46:53 PST 2018; root:xnu-4903.241.1~1/RELEASE_X86_64 x86_64`

**Description of the problem including expected versus actual behavior**: If the `.security` index is read-only, the application privileges API response does not identify the problematic index name in the response.

**Request**
``` json
PUT /_security/privilege/
{
  "myapp": {
    "read": {
      "actions": [
        "data:read/*",
        "action:login"
      ],
      "metadata": {
        "description": "Read access to myapp"
      }
    }
  }
}
```
**Response**
``` json
{
  "error": {
    "root_cause": [
      {
        "type": "cluster_block_exception",
        "reason": "blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];"
      }
    ],
    "type": "cluster_block_exception",
    "reason": "blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];"
  },
  "status": 403
}
```
This can be especially confusing for things like kibana logs:

`{"type":"log","@timestamp":"2019-03-23T02:01:15Z","tags":["security","error"],"pid":28144,"message":"Error registering Kibana Privileges with Elasticsearch for kibana-.kibana: [cluster_block_exception] blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];, with { suppressed={ 0={ type=\"cluster_block_exception\" & reason=\"blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];\" } & 1={ type=\"cluster_block_exception\" & reason=\"blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];\" } & 2={ type=\"cluster_block_exception\" & reason=\"blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];\" } } }"}`

**Steps to reproduce**:
 1. Set `.security` index to read only
    ``` json
    PUT .security/_settings
    {
        "index": {
        "blocks": {
            "read_only_allow_delete": "true"
            }
        }
    }
    ```
 2. Issue the `_security/privilege/` Request from above, and observe the response

### PR DESCRIPTION
**Pull request by codebird**
Updated the error message to reveal the index name that is causing it.

Closes #40870

### CODE REVIEW

**Comment by astefan** [comment]
Related to https://github.com/elastic/elasticsearch/issues/40870

**Comment by dnhatn** [comment]
@codebird Thanks so much for working on it.

Can you add a test verifies that ClusterBlockException contains the index name? You can add that new test to DeleteIndexBlocksIT.

**Comment by codebird** [comment]
Yep sure @dnhatn will work on that

**Comment by ielatif** [comment]
@dnhatn What about [indicesBlockedException(ClusterBlockLevel level, String[] indices)](https://github.com/elastic/elasticsearch/blob/106f59012de8f3b4932b53719f754612269fb04b/server/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java#L215) and [indicesAllowReleaseResources(String[] indices)](https://github.com/elastic/elasticsearch/blob/106f59012de8f3b4932b53719f754612269fb04b/server/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java#L238) ? Shouldn't add index name too ?

**Comment by codebird** [comment]
@cbuescher can you review this pls. maybe kick cans here too :)

**Comment by cbuescher** [comment]
Since @dnhatn is already on it and its more of his area I'm happy to leave it to him.

**Comment by codebird** [comment]
didn't mean anything bad for @dnhatn, I thought ultimately it should be reviewed from somebody inside the organization.

**Comment by codebird** [comment]
@dnhatn any more changes or advice here?

**Comment by dnhatn** [comment]
> What about indicesBlockedException(ClusterBlockLevel level, String[] indices) and indicesAllowReleaseResources(String[] indices) shouldn't add index name too ?

@ielatif Good find. Yes, we need to add index names for these cases too.

**Comment by dnhatn** [comment]
@codebird Thank you for picking this one up. Your approach looks great; however we need to make more changes to distinguish global blocks and index blocks; otherwise, the reporting message is still upsetting users. I drafted this change at https://github.com/dnhatn/elasticsearch/commit/e4af71b0d790c4fb9ca4c8c46b5879bc9b2d3166, can you adapt it in your PR?

**Comment by codebird** [comment]
@dnhatn  I adapted your code, but I reverted the work done on this function
`indicesAllowReleaseResources` because it was causing all DeleteIndexBlocksIT tests to fail,
the problem is that you are calling this `indicesBlockedException(ClusterBlockLevel.METADATA_WRITE, indices)` 
which doesn't filter by `clusterBlock.isAllowReleaseResources() == false`.

**Comment by codebird** [comment]
fixed this function now. Please check and let me know what you think. All tests pass, but for some reason I couldn't run the integration tests one last time (although they passed before), gradle is getting stuck on  `:build-tools:integTest`

**Comment by dnhatn** [comment]
> which doesn't filter by clusterBlock.isAllowReleaseResources()

@codebird  Good catch. I realized that my previous suggestion has a flaw where we don't return the blocking index name if only the global blocks exist. I made another adjustment based on your branch to fix this case. Can you please merge master into your branch and adapt the adjustment at https://github.com/dnhatn/elasticsearch/commit/82e7808aed4a786a7c8a23036b3cdab92ac03480? Thank you so much for working on it.

**Comment by codebird** [comment]
@dnhatn I'll modify the code a bit and keep the private function `getClusterAndIndexBlocks` but update it with your code just to remove duplication, because at the end `indicesBlockedException` and `indicesAllowReleaseResources` are doing the same thing but just filtering is different. I am running the tests now, but we still have the problem of this `DeleteIndexBlocksIT#testDeleteIndexOnIndexReadOnlyAllowDeleteSetting` failing. 

Should this `assertAcked(client().admin().indices().prepareDelete("test"));` be allowed when `INDEX_READ_ONLY_ALLOW_DELETE_BLOCK` is true?

**Comment by dnhatn** [comment]
> I'll modify the code a bit and keep the private function getClusterAndIndexBlocks but update it with your code just to remove duplication.

I prefer to keep these two method separately as they are pretty small.

> Should this assertAcked(client().admin().indices().prepareDelete("test")); be allowed when INDEX_READ_ONLY_ALLOW_DELETE_BLOCK is true?

We should not change this behaviour. The test `testDeleteIndexOnIndexReadOnlyAllowDeleteSetting` should pass now?

**Comment by codebird** [comment]
yep all tests pass. We're good to go.

**Comment by dnhatn** [comment]
@elasticmachine test this please.

**Comment by dnhatn** [comment]
@elasticmachine test this please.

**Comment by codebird** [comment]
@dnhatn I updated a test that failed, but I can't run all integration tests, gradle is getting stuck on this step `:build-tools:integTest` I am on ubuntu 18.04, do you have any idea how to get this resolved, I tried many things, but nothing seems to help

**Comment by dnhatn** [comment]
@elasticmachine test this please.

**Comment by codebird** [comment]
@dnhatn don't need to run it again. I'll fix my integTest, then recheck all the integration tests.

**Comment by codebird** [comment]
@dnhatn please check what I did, and run the tests if all good. I still can't get :build-tools:integTest to work :(

**Comment by dnhatn** [comment]
@elasticmachine  ok to test.

**Comment by codebird** [comment]
please run ci/1 again, its fail isn't related to this PR.

**Comment by dnhatn** [comment]
run elasticsearch-ci/1

**Comment by codebird** [comment]
again pls, not related.

File: server/src/test/java/org/elasticsearch/cluster/block/ClusterBlockTests.java [hunk 1]
```diff
@@ -109,7 +109,7 @@ public void testToStringDanglingComma() {
     public void testGlobalBlocksCheckedIfNoIndicesSpecified() {
         ClusterBlock globalBlock = randomClusterBlock();
         ClusterBlocks clusterBlocks = new ClusterBlocks(Collections.singleton(globalBlock), ImmutableOpenMap.of());
-        ClusterBlockException exception = clusterBlocks.indicesBlockedException(randomFrom(globalBlock.levels()), new String[0]);
+        ClusterBlockException exception = clusterBlocks.indicesBlockedException(randomFrom(globalBlock.levels()), new String[]{"test"});
```
Comment (dnhatn) [inline server/src/test/java/org/elasticsearch/cluster/block/ClusterBlockTests.java]: Can please revert this change?

File: server/src/test/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeActionTests.java [hunk 2]
```diff
@@ -287,7 +287,7 @@ public void testRequestBlock() {
             action.new AsyncAction(null, request, listener).start();
             fail("expected ClusterBlockException");
         } catch (ClusterBlockException expected) {
-            assertEquals("blocked by: [SERVICE_UNAVAILABLE/1/test-block];", expected.getMessage());
+            assertEquals("index ["+TEST_INDEX+"] blocked by: [SERVICE_UNAVAILABLE/1/test-block];", expected.getMessage());
```
Comment (dnhatn) [inline server/src/test/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeActionTests.java]: add spaces: `[" + TEST_INDEX + "]`

File: server/src/test/java/org/elasticsearch/action/admin/indices/delete/DeleteIndexBlocksIT.java [hunk 3]
```diff
@@ -63,7 +66,23 @@ public void testDeleteIndexOnIndexReadOnlyAllowDeleteSetting() {
...
+            createIndex("test");
+            ensureGreen("test");
+            Settings settings = Settings.builder().put(IndexMetaData.SETTING_READ_ONLY_ALLOW_DELETE, true).build();
+            client().admin().indices().prepareUpdateSettings("test").setSettings(settings).get();
+            ClusterBlockException e = expectThrows(ClusterBlockException.class, () ->
+                client().prepareIndex().setIndex("test").setType("doc").setId("1").setSource("foo", "bar").get());
+            assertEquals("index [test] blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];", e.getMessage());
+        } finally {
+            Settings settings = Settings.builder().putNull(IndexMetaData.SETTING_READ_ONLY_ALLOW_DELETE).build();
+            assertAcked(client().admin().indices().prepareUpdateSettings("test").setIndicesOptions(IndicesOptions.lenientExpandOpen()).
```
Comment (dnhatn) [inline server/src/test/java/org/elasticsearch/action/admin/indices/delete/DeleteIndexBlocksIT.java]: Maybe inline the settings and remove indices options.

File: server/src/test/java/org/elasticsearch/action/admin/indices/delete/DeleteIndexBlocksIT.java [hunk 4]
```diff
@@ -38,6 +39,7 @@ public void testDeleteIndexWithBlocks() {
             assertBlocked(client().admin().indices().prepareDelete("test"), MetaData.CLUSTER_READ_ONLY_BLOCK);
         } finally {
             setClusterReadOnly(false);
+            assertAcked(client().admin().indices().prepareDelete("test").get());
```
Comment (dnhatn) [inline server/src/test/java/org/elasticsearch/action/admin/indices/delete/DeleteIndexBlocksIT.java]: You don't have to add this manual delete.

**Review by dnhatn — CHANGES_REQUESTED** [review]
Thanks @codebird for working on it. This looks great already. Can you please address my comments and merge master into your branch.

File: server/src/test/java/org/elasticsearch/action/admin/indices/delete/DeleteIndexBlocksIT.java (hunk 4 above)
Comment (codebird) [reply server/src/test/java/org/elasticsearch/action/admin/indices/delete/DeleteIndexBlocksIT.java]: I was testing something. removing it now.

File: server/src/test/java/org/elasticsearch/action/admin/indices/delete/DeleteIndexBlocksIT.java [hunk 5]
```diff
@@ -53,7 +55,8 @@ public void testDeleteIndexOnIndexReadOnlyAllowDeleteSetting() {
             assertBlocked(client().prepareIndex().setIndex("test").setType("doc").setId("2").setSource("foo", "bar"),
                 IndexMetaData.INDEX_READ_ONLY_ALLOW_DELETE_BLOCK);
             assertBlocked(client().admin().indices().prepareUpdateSettings("test")
-                    .setSettings(Settings.builder().put("index.number_of_replicas", 2)), IndexMetaData.INDEX_READ_ONLY_ALLOW_DELETE_BLOCK);
+                    .setSettings(Settings.builder().put("index.number_of_replicas", 2)),
```
[the remark this answers is not in the document]
Comment (codebird) [reply server/src/test/java/org/elasticsearch/action/admin/indices/delete/DeleteIndexBlocksIT.java]: yep just line was too long, reverted.

File: server/src/test/java/org/elasticsearch/cluster/block/ClusterBlockTests.java (hunk 1 above)
Comment (codebird) [reply server/src/test/java/org/elasticsearch/cluster/block/ClusterBlockTests.java]: yep not needed anymore.

File: server/src/test/java/org/elasticsearch/action/admin/indices/delete/DeleteIndexBlocksIT.java (hunk 3 above)
Comment (codebird) [reply server/src/test/java/org/elasticsearch/action/admin/indices/delete/DeleteIndexBlocksIT.java]: just to be clear like this:

```assertAcked(client().admin().indices().prepareUpdateSettings("test")        .setSettings(Settings.builder().putNull(IndexMetaData.SETTING_READ_ONLY_ALLOW_DELETE).build()).get());```

Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/action/admin/indices/delete/DeleteIndexBlocksIT.java]: Yes

File: server/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java [hunk 6]
```diff
@@ -235,20 +228,27 @@ public ClusterBlockException indicesBlockedException(ClusterBlockLevel level, St
...
+            Set<ClusterBlock> blocks = Sets.union(globalBlocks, blocksForIndex(ClusterBlockLevel.METADATA_WRITE, index))
+                .stream().filter(clusterBlock -> clusterBlock.isAllowReleaseResources() == false).collect(toSet());
+            if (blocks.isEmpty() == false) {
+                indexLevelBlocks.put(index, Sets.union(globalBlocks, blocks));
+            }
+        }
+        if (indexLevelBlocks.isEmpty()) {
+            if(!globalBlocks.stream()
+                .filter(clusterBlock -> clusterBlock.isAllowReleaseResources() == false).collect(toSet()).isEmpty()){
+                return new ClusterBlockException(globalBlocks);
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java]: This is not correct. We should use the global blocks after filtered. Also, can we do this one at line 233?

File: server/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java [hunk 7]
```diff
@@ -199,34 +197,29 @@ public void indexBlockedRaiseException(ClusterBlockLevel level, String index) th
...
-            if (indexBlocked(level, index)) {
-                indexIsBlocked = true;
+            Set<ClusterBlock> indexBlocks = blocksForIndex(level, index);
+            if (indexBlocks.isEmpty() == false || globalLevelBlocks.isEmpty() == false) {
+                indexLevelBlocks.put(index, Sets.union(indexBlocks, globalLevelBlocks));
             }
         }
-        if (globalBlocked(level) == false && indexIsBlocked == false) {
+        if (indexLevelBlocks.isEmpty()) {
+            if(!globalLevelBlocks.isEmpty()){
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java]: Can we please change this to `globalLevelBlocks.isEmpty() == false` as we prefer "pred == false` which is easier to read than `!pred`.

**Review by dnhatn — COMMENTED** [review]
@codebird Thank you for the quick iteration. I left two more comments but your changes are very good.

File: server/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java (hunk 6 above)
Comment (codebird) [reply server/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java]: good catch

**Review by dnhatn — APPROVED** [review]
LGTM. 

Thanks @codebird.
